src/alignment/sparse.rs

```diff
@@ -0,0 +1,239 @@
+        events.push((x,y,(idx+matches.len()) as u32));
+        events.push((x+k,y+k,idx as u32));
+
+        n = max(n, x+k);
+        n = max(n, y+k);
+    }
+    events.sort(); 
+
+
+    let mut max_col_dp: MaxBitTree<(u32, u32)> = MaxBitTree::new(n as usize);
+    let mut dp: Vec<(u32, i32)> = Vec::with_capacity(events.len());
+    let mut best_dp = (0,-1);
+
+    for _ in 0 .. events.len() {
+        dp.push((0,0));
+    }
+
+    for ev in events {
+        let p = (ev.2 % matches.len() as u32) as usize;
+        let j = ev.1;
+        let is_start = ev.2 >= (matches.len() as u32);
+        
+
+        if is_start {
+            dp[p] = (k, -1);
+            let (best_value, best_position) = max_col_dp.get(j as usize);
+            if best_value > 0 {
+                dp[p] = (k + best_value, best_position as i32);
+                best_dp = max(best_dp, (dp[p].0, p as i32));
+            }
+        } else {
+            // See if this kmer continues a diffent kmer
+            if ev.0 >= k+1 && ev.1 >= k+1 {
+                match matches.binary_search(&(ev.0-k-1, ev.1-k-1)) {
+                    Ok(cont_idx) => {
+                        let prev_score = dp[cont_idx].0;
+                        let candidate = (prev_score + 1, cont_idx as i32);
+                        dp[p] = max(dp[p], candidate);
+                        best_dp = max(best_dp, (dp[p].0, p as i32));
+                    },
+                    _ => (),
+                }
+            }
+
+            max_col_dp.set(ev.1 as usize, (dp[p].0, p as u32));
+        }
+    }
+
+    let mut traceback = Vec::new();
+    let (best_score, mut prev_match) = best_dp;
+    while prev_match >= 0 {
+        traceback.push(prev_match as usize);
+        prev_match = dp[prev_match as usize].1;
+    }
+    traceback.reverse();
+    SparseAlignmentResult { path: traceback, score: best_score, dp_vector: dp }
+}
+
+
+/// Find all matches of length k between two strings, using an efficient q-gram
+/// index. For very long reference strings, it may be more efficient to use and
+/// FMD index to generate the matches. Note that this method is mainly for 
+/// demonstration & testing purposes.  For aligning many query sequences
+/// against the same reference, you should reuse the QGramIndex of the reference.
+pub fn find_kmer_matches<T: AsRef<[u8]>>(query: &T, reference: &T, k: usize) -> Vec<(u32, u32)> {
+
+    let slc1 = query.as_ref();
+    let slc2 = reference.as_ref();
+
+    let mut alphabet = Alphabet::new(slc2);
+    if !alphabet.is_word(slc1) {
+        for i in slc1 {
+            alphabet.insert(*i)
+        }
+    }
+
+    let qgram = QGramIndex::new(k as u32, slc2, &alphabet);
+    let qg_matches = qgram.exact_matches(slc1);
+
+    let mut matches = Vec::new();
+
+    // Q-gram matches will be >= k in length, but each individual kmer match 
+    // must appear in mathces, so for a Q-gram match of length L, 
+    // generate L - k + 1 matches.hh
```

Comment:
I see. But this is kind of redundant. You can simply use the method `qgram_matches` for every q/gram of the query string. This will be also faster.